### ISSUE TITLE
Need better scheme for "method contexts"

### ISSUE THREAD

**Issue by yangsophieee** [traitecoevo/traits.build#16]
Referencing this [issue from austraits.build](https://github.com/traitecoevo/austraits.build/issues/654):

Currently, there are instances where the same trait is measured in the same dataset using multiple methods. These instances all have a `method_context` assigned that indicates the differences between the two methods. This `method_context` is captured in `austraits$methods`, `austraits$contexts` and `austraits$traits` (`method_id`). In `austraits$methods` there are still 2 or more entries for a given `trait_name` x `dataset_id` combination with no indication of which row goes with which measurements. As a result, when you merge the methods table into the trait table, the actual trait measurements for those `trait_name` x `dataset_id` combinations are being duplicated, with each measurement assigned all methods.

For now we are going to standardise the methods entered and move more information to method context. This solves the duplication, but isn't a good long term fix (Lizzy's words -- I'm not sure how this solves duplication). The current fix is not really satisfactory - not least, because it would be easy to input two traits with different methods and not realise why rows of data are being duplicated when you merge in the methods table.

A better solution is to separate `method_context` from the other contextual properties, since method context is a specialisation of method, not a context of an observation. Method context would be better mapped through the methods table, but this is a much bigger fix.

The following studies currently have much of their methods moved to method contexts for traits that have 2 entries, to avoid duplication:
- Pickup_2005
- Ghannoum_2010
- Wright_2006
- Wright_2019
- Kotowska_2020

**Comment by yangsophieee** [comment]
This is @ehwenk's code, not quite finished, that adds `method_id`'s to the methods table, although only for those method contexts that are generated in metadata[["traits"]]. It is clunky, but a possible direction to expand upon.
```
  # identify sources as being `primary`, `secondary` or `original`
  # secondary datasets are additional publications associated with the primary citation
  # original dataset keys are used for compilations indicating the original data sources
  citation_types <- 
    tibble::tibble(
      source_key = names(metadata$source),
      type = str_replace_all(.data$source_key, "_[:digit:]+", ""),
      source_id = metadata$source %>%
        util_list_to_df2() %>%
        dplyr::select(.data$key)
    ) 

  source_primary_key <- metadata$source$primary$key
  source_secondary_keys <- citation_types %>% dplyr::filter(.data$type == "secondary") %>% dplyr::select(.data$source_id) %>% as.vector() 
  source_secondary_keys <- source_secondary_keys$source_id$key %>% as.vector()
  source_original_dataset_keys <- citation_types %>% dplyr::filter(.data$type == "original") %>% dplyr::select(.data$source_id) %>% as.vector()
  source_original_dataset_keys <- source_original_dataset_keys$source_id$key %>% as.vector()
  
  # combine collectors to add into the methods table
  collectors_tmp <-
    stringr::str_c(contributors$given_name, " ",
                   contributors$last_name,
                   ifelse(!is.na(contributors$additional_role),
                          paste0(" (", contributors$additional_role, ")"),
                          ""))  %>% paste(collapse = ", ")

  # contexts that are of category = method
 # browser()
  if(nrow(contexts) > 0) {
     if(nrow(contexts %>% filter(category == "method")) > 0) {
    method_contexts_tmp <-
      contexts %>%
        filter(category == "method") %>% 
        distinct(var_in)
     } else {
       method_contexts_tmp <- tibble::tibble(var_in = "XX")
     }
  } else {
    method_contexts_tmp <- tibble::tibble(var_in = "XX")
  }
  
  # identify which method contexts come from metadata[["traits"]]
  # XXXX But I know it can't just be searching for method_contexts_tmp$var_in[[1]] - needs a seq_along, but this is a start
  
  if(nrow(contexts) > 0) { 
    traits_columns <-  metadata[["traits"]] %>%
      util_list_to_df2() %>% dplyr::select(dplyr::any_of(method_contexts_tmp$var_in[[1]]))
  } else {
    traits_columns <- tibble::tibble()
  }
    
  # TRUE/FALSE variable, which is TRUE if there are method contexts keyed in through metadata[["traits]] and FALSE if all metadata contexts come from columns
  # if FALSE, then there is no need for those method_ids to be in the methods tables and in fact they can't be in the methods table, because they are likely
  # to vary across rows of data, not simply across trait entries
  
  traits_columns_tmp <- ifelse(ncol(traits_columns) == 0 | is.null(traits_columns), FALSE, TRUE)
  
  #browser()
  
  methods <-
    dplyr::full_join( by = "dataset_id",
      # methods used to collect each trait
      metadata[["traits"]] %>%
        util_list_to_df2() %>%
        dplyr::filter(!is.na(.data$trait_name)) %>%
        dplyr::mutate(dataset_id = dataset_id) %>%
  #      dplyr::select(dataset_id, .data$trait_name, .data$methods, dplyr::any_of("method_context"))
        dplyr::select(dataset_id, .data$trait_name, .data$methods, dplyr::any_of(method_contexts_tmp$var_in[[1]]))
      ,
      # study methods
      metadata$dataset %>%
        util_list_to_df1() %>%
        tidyr::spread(.data$key, .data$value) %>%
        dplyr::select(dplyr::any_of(names(metadata$dataset))) %>%
        dplyr::mutate(dataset_id = dataset_id) %>%
        dplyr::select(-dplyr::any_of(c("original_file", "notes", "data_is_long_format", "taxon_name", 
                                         "trait_name", "population_id", "individual_id",
                                         "location_name", "source_id", "value", "entity_type", 
                                         "collection_date", "custom_R_code", "replicates", "measurement_remarks",
                                         "taxon_name", "basis_of_value", "basis_of_record", "life_stage")))
      )  %>%
      full_join( by = "dataset_id",
      # references
        tibble::tibble(
          dataset_id = dataset_id,
          source_primary_key = source_primary_key,
          source_primary_citation = bib_print(sources[[source_primary_key]]),
          source_secondary_key = source_secondary_keys %>% paste(collapse = "; "),
          source_secondary_citation = ifelse(length(source_secondary_keys) == 0, NA_character_,
            purrr::map_chr(sources[source_secondary_keys], bib_print) %>% paste(collapse = "; ") %>%
              stringr::str_replace_all("\\.;", ";")
            ),                    
          source_original_dataset_key = source_original_dataset_keys %>% paste(collapse = "; "),
          source_original_dataset_citation = ifelse(length(source_original_dataset_keys) == 0, NA_character_,
            purrr::map_chr(sources[source_original_dataset_keys], bib_print) %>% paste(collapse = "; ") %>%
            stringr::str_replace_all("\\.;", ";")
          )
        )
      ) %>%
    dplyr::mutate(
      data_collectors = collectors_tmp,
      assistants = ifelse(is.null(metadata$contributors$assistants), NA_character_,
                                      metadata$contributors$assistants),
      austraits_curators = metadata$contributors$austraits_curators
    )


  method_contexts <- 
    context_ids$contexts %>%
      filter(category == "method", link_id == "method_id", var_in %in% traits_columns) %>%
      rename(method_context = value, method_id = link_vals) %>%
      select(method_context, method_id)

  if(nrow(method_contexts) > 0 ) {
    methods <-
      methods %>%
      left_join(by = "method_context", method_contexts) %>% 
      select(c("dataset_id", "trait_name", "method_id", "method_context", "methods"), everything())
  } else {
    methods <-
      methods %>%
      mutate(method_id = NA_character_)
  }

  methods %>% 
  select(-any_of(c("method_context"))) %>%
  select(c("dataset_id", "trait_name", "method_id", "methods"), everything())
}
```

**Comment by yangsophieee** [comment]
Discussed this with @ehwenk today and we decided it would be pretty straightforward to add an extra ID for the methods. Originally we thought we could just run `distinct` on the methods table to prevent duplication when `left_join`'ing to the traits table, but there are cases when a study has two of the same trait with different methods, not just different method contexts, which would still lead to duplication. 

It should be easy to add a new ID called `method_id` to identify unique methods from the methods table, and change the name of the current method context ID (`method_id`) to `method_context_id` and keep `method_context_id` in the contexts table.

If this can be done, it'll be a really great breakthrough as this has been causing duplication issues for a while!

**Comment by ehwenk** [comment]
Few more tweaks and edits to all necessary metdata files in austraits.build to ensure this works - it does.
(see commit https://github.com/traitecoevo/traits.build/commit/e9643c235253daf936213b8ae3d2526e2c34a4eb)

> austraits$methods %>% nrow()
[1] 3540
> austraits$traits %>% nrow()
[1] 1800492
> austraits$traits %>% left_join(austraits$methods, by = c("dataset_id", "trait_name", "method_id")) %>% nrow()
[1] 1800492

And this is the same number of rows you get when building austraits from the develop branch of traits.build.

**Comment by ehwenk** [comment]
@yangsophieee 

Can you write a test that checks for this:
- add or manipulate a test dataset, such that there are multiple entries for a single traits under metadata[["traits"]], with a mix of different value types, and methods
- different method fields result in automatically added method_id's > 1
- different value_types not considered (i.e. method_id remains 1)
- repeats of a trait with the same method field not considered (i.e. method_id remains 1)

**Comment by ehwenk** [comment]
closed with [8714247](https://github.com/traitecoevo/traits.build/commit/8714247b22744216c23d2c533f277b576b5d5d5f)
